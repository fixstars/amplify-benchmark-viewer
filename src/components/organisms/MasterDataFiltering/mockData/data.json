[
  { "name": "Tsp" },
  { "name": "Qap" },
  { "name": "MaxCut" },
  { "name": "Sudoku" },
  { "name": "Cvrp" }
]
