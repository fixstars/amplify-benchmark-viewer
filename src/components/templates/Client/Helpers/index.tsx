import type { GridColumns } from '@mui/x-data-grid'
import { Link } from 'react-router-dom'

export interface ClientParameter {
  readonly [key: string]:
    | string
    | boolean
    | number
    | ClientParameter
    | undefined
}

export interface ClientData {
  readonly class: string
  readonly instance: string
  readonly problemParameters: ClientParameter
  readonly version: string
  readonly '0%': number | null
  readonly '1%': number | null
  readonly '5%': number | null
  readonly '10%': number | null
  readonly '20%': number | null
  readonly '50%': number | null
}

const renderInstanceCell = (row: ClientData) => {
  return (
    <Link to={`/classes/${row.class}/instances/${row.instance}`}>
      {row.instance}
    </Link>
  )
}

const renderProblemParametersCell = ({
  problemParameters,
  ...client
}: ClientData) => {
  return Object.entries(problemParameters).map(
    ([parameterName, parameterValue]) =>
      typeof parameterValue === 'object' ? (
        <div
          key={`${client.class}_${client.instance}_${client.version}_${parameterName}`}
        >
          <div>{parameterName}</div>
          {Object.entries(parameterValue).map(([key, value]) => (
            <div
              key={`${parameterName}_${key}`}
              style={{ paddingLeft: 10 }}
            >{`${key}: ${value?.toString()}`}</div>
          ))}
        </div>
      ) : (
        <div
          key={`${client.class}_${client.instance}_${client.version}_${parameterName}_${parameterValue}`}
        >
          {parameterName}: {parameterValue?.toString()}
        </div>
      ),
  )
}

const renderTTSCell = (value: number | null) => {
  return value == null ? '-' : value.toString()
}

const columns: GridColumns<ClientData> = [
  {
    field: 'class',
    headerName: 'Class',
    sortable: false,
  },
  {
    field: 'instance',
    headerName: 'Instance',
    sortable: false,
    minWidth: 140,
    renderCell: (params) => renderInstanceCell(params.row),
  },
  {
    field: 'problemParameters',
    headerName: 'Problem parameters',
    sortable: false,
    cellClassName: 'problemParameters',
    minWidth: 180,
    renderCell: (params) => renderProblemParametersCell(params.row),
  },
  {
    field: 'version',
    headerName: 'Version',
    sortable: false,
    minWidth: 220,
  },
  {
    field: '0%',
    headerName: 'TTS(0%)',
    sortable: false,
    minWidth: 160,
    renderCell: (params) => renderTTSCell(params.row['0%']),
  },
  {
    field: '1%',
    headerName: 'TTS(1%)',
    sortable: false,
    minWidth: 160,
    renderCell: (params) => renderTTSCell(params.row['1%']),
  },
  {
    field: '5%',
    headerName: 'TTS(5%)',
    sortable: false,
    minWidth: 160,
    renderCell: (params) => renderTTSCell(params.row['5%']),
  },
  {
    field: '10%',
    headerName: 'TTS(10%)',
    sortable: false,
    minWidth: 160,
    renderCell: (params) => renderTTSCell(params.row['10%']),
  },
  {
    field: '20%',
    headerName: 'TTS(20%)',
    sortable: false,
    minWidth: 160,
    renderCell: (params) => renderTTSCell(params.row['20%']),
  },
  {
    field: '50%',
    headerName: 'TTS(50%)',
    sortable: false,
    minWidth: 160,
    renderCell: (params) => renderTTSCell(params.row['50%']),
  },
]

export { columns }
