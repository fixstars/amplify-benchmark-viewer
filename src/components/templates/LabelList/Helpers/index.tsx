/**
 * Copyright (c) Fixstars Corporation and Fixstars Amplify Corporation.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type { GridColDef } from '@mui/x-data-grid'
import type { ProblemInstances } from 'utils/DataGridCell'
import {
  renderClientsCell,
  renderProblemInstancesCell,
} from 'utils/DataGridCell'

export interface LabelData {
  readonly label: string
  readonly problemInstances: ProblemInstances
  readonly clients: ReadonlyArray<string>
}

// eslint-disable-next-line functional/prefer-readonly-type
const columns: GridColDef<LabelData>[] = [
  {
    field: 'label',
    headerName: 'Label',
    sortable: false,
    minWidth: 220,
  },
  {
    field: 'instance',
    headerName: 'Instance',
    sortable: false,
    flex: 1,
    cellClassName: 'instance',
    renderCell: ({ row: { label, problemInstances } }) =>
      renderProblemInstancesCell(label, problemInstances),
  },
  {
    field: 'clients',
    headerName: 'Clients',
    sortable: false,
    flex: 1,
    renderCell: ({ row }) => renderClientsCell(row.label, row.clients),
  },
]

export { columns }
