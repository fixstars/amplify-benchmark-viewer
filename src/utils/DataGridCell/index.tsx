/**
 * Copyright (c) Fixstars Corporation and Fixstars Amplify Corporation.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { LinkButton } from 'components/atoms/LinkButton'

const clientColors = [
  '#0f66dc',
  '#f57bae',
  '#e75e75',
  '#e5f97c',
  '#1b000d',
  '#f6a695',
  '#692981',
  '#3cde29',
  '#591414',
  '#d66594',
  '#2d884a',
  '#400763',
  '#9d19f4',
  '#fcdeb9',
  '#1f4c48',
  '#909743',
  '#2dc1c8',
  '#c17913',
  '#b2a1f6',
]

const instanceColors = [
  '#ff66ff',
  '#0f66dc',
  '#f57bae',
  '#e75e75',
  '#e5f97c',
  '#1b000d',
  '#f6a695',
  '#692981',
  '#3cde29',
  '#591414',
  '#d66594',
  '#2d884a',
  '#400763',
  '#9d19f4',
  '#fcdeb9',
  '#718de6',
  '#1f4c48',
  '#909743',
  '#2dc1c8',
  '#c17913',
  '#b2a1f6',
  '#9a916d',
  '#406a0f',
  '#004b31',
  '#18729f',
  '#9997c3',
  '#9cf0ec',
  '#ce843c',
  '#2d326e',
  '#76ed60',
  '#d7f7e8',
  '#bb15d7',
  '#87a7d8',
  '#94053a',
  '#e5c8a6',
  '#e2f2c0',
]

export interface ProblemInstances {
  readonly [key: string]: ReadonlyArray<string> | undefined
}

const renderClientsCell = (key: string, clients: ReadonlyArray<string>) => {
  const getBackgroundColor = (client: string): string => {
    const cache = JSON.parse(localStorage.getItem('clientColorCache') ?? '{}')
    let cacheIndex = JSON.parse(
      localStorage.getItem('clientColorIndexCache') ?? '-1',
    )
    let backgroundColor = cache[client]

    if (backgroundColor === undefined) {
      const length = clientColors.length
      cacheIndex = cacheIndex >= length ? 0 : cacheIndex + 1
      backgroundColor = clientColors[cacheIndex]
      cache[client] = backgroundColor
      localStorage.setItem('clientColorCache', JSON.stringify(cache))
      localStorage.setItem('clientColorIndexCache', JSON.stringify(cacheIndex))
      return backgroundColor
    }

    return backgroundColor
  }

  return clients.map((client) => (
    <div key={`${key}_${client}`} style={{ marginRight: 10 }}>
      <LinkButton
        label={client}
        link={`/clients/${client}`}
        backgroundColor={getBackgroundColor(client)}
      />
    </div>
  ))
}

const renderProblemInstancesCell = (
  key: string,
  problemInstances: ProblemInstances,
) => {
  const getBackgroundColor = (instance: string): string => {
    const cache = JSON.parse(localStorage.getItem('instanceColorCache') ?? '{}')
    let cacheIndex = JSON.parse(
      localStorage.getItem('instanceColorIndexCache') ?? '-1',
    )
    let backgroundColor = cache[instance]

    if (backgroundColor === undefined) {
      const length = instanceColors.length
      cacheIndex = cacheIndex >= length ? 0 : cacheIndex + 1
      backgroundColor = instanceColors[cacheIndex]
      cache[instance] = backgroundColor
      localStorage.setItem('instanceColorCache', JSON.stringify(cache))
      localStorage.setItem(
        'instanceColorIndexCache',
        JSON.stringify(cacheIndex),
      )
      return backgroundColor
    }

    return backgroundColor
  }
  const pairs = Object.entries(problemInstances)
  pairs.sort(function (p1, p2) {
    const p1Key = p1[0]
    const p2Key = p2[0]
    if (p1Key < p2Key) {
      return -1
    }
    if (p1Key > p2Key) {
      return 1
    }
    return 0
  })
  problemInstances = Object.fromEntries(pairs)
  return Object.keys(problemInstances).map((className) => (
    <div
      key={`${key}_${className}`}
      style={{ margin: '5px 0 5px 0', display: 'flex' }}
    >
      <span style={{ marginRight: 10 }}>{className}:</span>
      <div style={{ display: 'flex', flexWrap: 'wrap' }}>
        {problemInstances[className]?.map((instance) => (
          <div
            key={`${key}_${className}_${instance}`}
            style={{ marginRight: 10, marginBottom: 10 }}
          >
            <LinkButton
              label={instance}
              link={`/classes/${className}/instances/${instance}`}
              backgroundColor={getBackgroundColor(instance)}
            />
          </div>
        ))}
      </div>
    </div>
  ))
}

export { renderClientsCell, renderProblemInstancesCell }
