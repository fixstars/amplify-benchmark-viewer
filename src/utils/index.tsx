export * from './DataGridCell'
export * from './statistics'
