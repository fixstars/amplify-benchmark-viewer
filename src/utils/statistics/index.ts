function quartile(data: readonly number[], q: number) {
  const sortedData = data.slice().sort((a, b) => a - b)
  const pos = Math.floor((sortedData.length - 1) * q)
  return sortedData[pos]
}
function quartile0(data: readonly number[]) {
  return quartile(data, 0)
}
function quartile25(data: readonly number[]) {
  return quartile(data, 0.25)
}

function quartile50(data: readonly number[]) {
  return quartile(data, 0.5)
}

function quartile75(data: readonly number[]) {
  return quartile(data, 0.75)
}

function quartile100(data: readonly number[]) {
  return quartile(data, 1)
}

export { quartile0, quartile25, quartile50, quartile75, quartile100 }
