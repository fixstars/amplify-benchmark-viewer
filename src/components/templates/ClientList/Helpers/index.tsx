/**
 * Copyright (c) Fixstars Corporation and Fixstars Amplify Corporation.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type { GridColumns } from '@mui/x-data-grid'
import { Link } from 'react-router-dom'
import type { ProblemInstances } from 'utils/DataGridCell'
import { renderProblemInstancesCell } from 'utils/DataGridCell'

export interface ClientInstances {
  readonly client: string
  readonly version: string
  readonly problemInstances: ProblemInstances
}

const renderClientCell = (row: ClientInstances) => {
  return <Link to={`/clients/${row.client}`}>{row.client}</Link>
}

const columns: GridColumns<ClientInstances> = [
  {
    field: 'client',
    headerName: 'Client',
    sortable: false,
    minWidth: 180,
    renderCell: (params) => renderClientCell(params.row),
  },
  {
    field: 'version',
    headerName: 'Version',
    sortable: false,
    minWidth: 220,
  },
  {
    field: 'instance',
    headerName: 'Instance',
    sortable: false,
    flex: 1,
    cellClassName: 'instance',
    renderCell: ({ row: { client, version, problemInstances } }) =>
      renderProblemInstancesCell(`${client}_${version}`, problemInstances),
  },
]

export { columns }
