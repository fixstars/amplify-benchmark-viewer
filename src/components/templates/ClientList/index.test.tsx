import { render } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import { mockLayout } from 'utils/test'
import mockData from 'utils/test/mocks/data/mockClientListData.json'

import { ClientList } from '.'
import { columns } from './Helpers'

describe('<ClientList />', () => {
  it('Rendered well', async () => {
    const { container } = render(
      <BrowserRouter>
        <ClientList data={mockData} />
      </BrowserRouter>,
    )

    const layout = mockLayout.mock.calls[0][0]
    expect(layout.children.length).toBe(2)

    const sectionTitleContainer = layout.children[0]
    expect(sectionTitleContainer.type.name).toBe('Box')
    const sectionTitle = sectionTitleContainer.props.children
    expect(sectionTitle.type.name).toBe('SectionTitle')
    expect(sectionTitle.props.title).toBe('Client List')

    const contentsContainer = layout.children[1]
    expect(contentsContainer.type.name).toBe('Box')
    expect(contentsContainer.props.sx).toEqual({ display: 'flex', flex: 1 })
    const dataGrid = contentsContainer.props.children
    expect(dataGrid.type.name).toBe('DataGrid')
    expect(dataGrid.props.sx).toEqual({
      '&.MuiDataGrid-root .MuiDataGrid-cell:focus-within': {
        outline: 'none !important',
      },
      '&.MuiDataGrid-root .MuiDataGrid-cell.instance': {
        alignItems: 'flex-start',
        flexDirection: 'column',
        padding: '5px 0 5px 0',
      },
    })
    expect(dataGrid.props.rows).toEqual(mockData)
    expect(dataGrid.props.columns).toEqual(columns)
    expect(dataGrid.props.getRowId.name).toBe('getRowId')
    expect(dataGrid.props.hideFooter).toBe(true)
    expect(dataGrid.props.disableColumnMenu).toBe(true)

    expect(container).toMatchSnapshot()
  })
})
