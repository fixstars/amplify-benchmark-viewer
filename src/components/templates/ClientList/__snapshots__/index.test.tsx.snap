// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClientList /> Rendered well 1`] = `
<div>
  <div
    class="MuiBox-root css-thz6x0"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-1hc7nu0-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1r339ym-MuiToolbar-root"
      >
        <div
          class="css-yp9swi"
        >
          <a
            class="css-94l1l1"
            href="/"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-19sbe58-MuiTypography-root"
            >
              Annealing Benchmark
            </h6>
          </a>
        </div>
        <nav
          class="css-ov1ktg"
        >
          <a
            class="css-94l1l1"
            href="/"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-4hrd4f-MuiTypography-root"
            >
              Problems
            </p>
          </a>
          <a
            class="css-94l1l1"
            href="/clients"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-muv6ew-MuiTypography-root"
            >
              Clients
            </p>
          </a>
          <a
            class="css-94l1l1"
            href="/labels"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-muv6ew-MuiTypography-root"
            >
              Label
            </p>
          </a>
        </nav>
      </div>
    </header>
    <div
      class="MuiBox-root css-1cf7osc"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
      />
      <div
        class="MuiBox-root css-0"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-15h1z56-MuiGrid-root"
        >
          <h2
            class="MuiTypography-root MuiTypography-h2 css-1asg1ln-MuiTypography-root"
          >
            Client List
          </h2>
        </div>
      </div>
      <div
        class="MuiBox-root css-13pmxen"
      >
        <div
          aria-colcount="3"
          aria-multiselectable="false"
          aria-rowcount="4"
          class="MuiDataGrid-root MuiDataGrid-root--densityStandard css-157vse1-MuiDataGrid-root"
          role="grid"
        >
          <div>
            <div />
          </div>
          <div
            class="MuiDataGrid-main css-204u17-MuiDataGrid-main"
          >
            <div
              class="MuiDataGrid-columnHeaders css-f3jnds-MuiDataGrid-columnHeaders"
              style="min-height: 56px; max-height: 56px; line-height: 56px;"
            >
              <div
                class="MuiDataGrid-columnHeadersInner MuiDataGrid-columnHeadersInner--scrollable css-gl260s-MuiDataGrid-columnHeadersInner"
                role="rowgroup"
                style="transform: translate3d(0px, 0px, 0px);"
              >
                <div
                  aria-rowindex="1"
                  class="css-yrdy0g-MuiDataGrid-columnHeaderRow"
                  role="row"
                >
                  <div
                    aria-colindex="1"
                    aria-label="Client"
                    aria-sort="none"
                    class="MuiDataGrid-columnHeader"
                    data-field="client"
                    role="columnheader"
                    style="height: 56px; width: 140px; min-width: 140px; max-width: 140px;"
                    tabindex="0"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderDraggableContainer"
                      draggable="false"
                    >
                      <div
                        class="MuiDataGrid-columnHeaderTitleContainer"
                      >
                        <div
                          class="MuiDataGrid-columnHeaderTitleContainerContent"
                        >
                          <div
                            aria-label=""
                            class="MuiDataGrid-columnHeaderTitle css-1jbbcbn-MuiDataGrid-columnHeaderTitle"
                            data-mui-internal-clone-element="true"
                          >
                            Client
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                      style="min-height: 56px; opacity: 1;"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                        data-testid="SeparatorIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M11 19V5h2v14z"
                        />
                      </svg>
                    </div>
                  </div>
                  <div
                    aria-colindex="2"
                    aria-label="Version"
                    aria-sort="none"
                    class="MuiDataGrid-columnHeader"
                    data-field="version"
                    role="columnheader"
                    style="height: 56px; width: 220px; min-width: 220px; max-width: 220px;"
                    tabindex="-1"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderDraggableContainer"
                      draggable="false"
                    >
                      <div
                        class="MuiDataGrid-columnHeaderTitleContainer"
                      >
                        <div
                          class="MuiDataGrid-columnHeaderTitleContainerContent"
                        >
                          <div
                            aria-label=""
                            class="MuiDataGrid-columnHeaderTitle css-1jbbcbn-MuiDataGrid-columnHeaderTitle"
                            data-mui-internal-clone-element="true"
                          >
                            Version
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                      style="min-height: 56px; opacity: 1;"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                        data-testid="SeparatorIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M11 19V5h2v14z"
                        />
                      </svg>
                    </div>
                  </div>
                  <div
                    aria-colindex="3"
                    aria-label="Instance"
                    aria-sort="none"
                    class="MuiDataGrid-columnHeader"
                    data-field="instance"
                    role="columnheader"
                    style="height: 56px; width: 0px; min-width: 0; max-width: 0;"
                    tabindex="-1"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderDraggableContainer"
                      draggable="false"
                    >
                      <div
                        class="MuiDataGrid-columnHeaderTitleContainer"
                      >
                        <div
                          class="MuiDataGrid-columnHeaderTitleContainerContent"
                        >
                          <div
                            aria-label=""
                            class="MuiDataGrid-columnHeaderTitle css-1jbbcbn-MuiDataGrid-columnHeaderTitle"
                            data-mui-internal-clone-element="true"
                          >
                            Instance
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                      style="min-height: 56px; opacity: 1;"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                        data-testid="SeparatorIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M11 19V5h2v14z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              style="overflow: visible; height: 0px; width: 0px;"
            >
              <div
                class="MuiDataGrid-virtualScroller css-1w5m2wr-MuiDataGrid-virtualScroller"
                style="width: 0px; height: auto; margin-top: 56px; overflow-x: hidden;"
              >
                <div
                  class="MuiDataGrid-virtualScrollerContent MuiDataGrid-virtualScrollerContent--overflowed css-1kwdphh-MuiDataGrid-virtualScrollerContent"
                  style="width: auto; height: 156px; min-height: auto;"
                >
                  <div
                    class="MuiDataGrid-virtualScrollerRenderZone css-s1v7zr-MuiDataGrid-virtualScrollerRenderZone"
                    style="transform: translate3d(0px, 0px, 0px);"
                  >
                    <div
                      aria-rowindex="2"
                      aria-selected="false"
                      class="MuiDataGrid-row MuiDataGrid-row--dynamicHeight"
                      data-id="a_client_v0.6.4-54-ge56964d-V100"
                      data-rowindex="0"
                      role="row"
                      style="max-height: none; min-height: auto;"
                    >
                      <div
                        aria-colindex="1"
                        aria-colspan="1"
                        class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                        data-colindex="0"
                        data-field="client"
                        role="cell"
                        style="min-width: 140px; max-width: 140px; min-height: auto; max-height: none;"
                        tabindex="-1"
                      >
                        <a
                          href="/clients/a_client"
                        >
                          a_client
                        </a>
                      </div>
                      <div
                        aria-colindex="2"
                        aria-colspan="1"
                        class="MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                        data-colindex="1"
                        data-field="version"
                        role="cell"
                        style="min-width: 220px; max-width: 220px; min-height: auto; max-height: none;"
                        tabindex="-1"
                      >
                        <div
                          class="MuiDataGrid-cellContent"
                        >
                          v0.6.4-54-ge56964d-V100
                        </div>
                      </div>
                      <div
                        aria-colindex="3"
                        aria-colspan="1"
                        class="instance MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                        data-colindex="2"
                        data-field="instance"
                        role="cell"
                        style="min-width: 0; max-width: 0; min-height: auto; max-height: none;"
                        tabindex="-1"
                      >
                        <div
                          style="margin: 5px 0px 5px 0px; display: flex;"
                        >
                          <span
                            style="margin-right: 10px;"
                          >
                            MaxCut
                            :
                          </span>
                          <div
                            style="margin-right: 10px;"
                          >
                            <a
                              class="css-1o4kjj1"
                              href="/classes/MaxCut/instances/d_instance"
                            >
                              d_instance
                            </a>
                          </div>
                          <div
                            style="margin-right: 10px;"
                          >
                            <a
                              class="css-1o7gueb"
                              href="/classes/MaxCut/instances/e_instance"
                            >
                              e_instance
                            </a>
                          </div>
                          <div
                            style="margin-right: 10px;"
                          >
                            <a
                              class="css-gqb4bp"
                              href="/classes/MaxCut/instances/f_instance"
                            >
                              f_instance
                            </a>
                          </div>
                        </div>
                        <div
                          style="margin: 5px 0px 5px 0px; display: flex;"
                        >
                          <span
                            style="margin-right: 10px;"
                          >
                            Tsp
                            :
                          </span>
                          <div
                            style="margin-right: 10px;"
                          >
                            <a
                              class="css-196ffja"
                              href="/classes/Tsp/instances/a_instance"
                            >
                              a_instance
                            </a>
                          </div>
                          <div
                            style="margin-right: 10px;"
                          >
                            <a
                              class="css-14frf9o"
                              href="/classes/Tsp/instances/b_instance"
                            >
                              b_instance
                            </a>
                          </div>
                          <div
                            style="margin-right: 10px;"
                          >
                            <a
                              class="css-tudjox"
                              href="/classes/Tsp/instances/c_instance"
                            >
                              c_instance
                            </a>
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      aria-rowindex="3"
                      aria-selected="false"
                      class="MuiDataGrid-row MuiDataGrid-row--dynamicHeight"
                      data-id="a_client_v0.7.0-V100"
                      data-rowindex="1"
                      role="row"
                      style="max-height: none; min-height: auto;"
                    >
                      <div
                        aria-colindex="1"
                        aria-colspan="1"
                        class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                        data-colindex="0"
                        data-field="client"
                        role="cell"
                        style="min-width: 140px; max-width: 140px; min-height: auto; max-height: none;"
                        tabindex="-1"
                      >
                        <a
                          href="/clients/a_client"
                        >
                          a_client
                        </a>
                      </div>
                      <div
                        aria-colindex="2"
                        aria-colspan="1"
                        class="MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                        data-colindex="1"
                        data-field="version"
                        role="cell"
                        style="min-width: 220px; max-width: 220px; min-height: auto; max-height: none;"
                        tabindex="-1"
                      >
                        <div
                          class="MuiDataGrid-cellContent"
                        >
                          v0.7.0-V100
                        </div>
                      </div>
                      <div
                        aria-colindex="3"
                        aria-colspan="1"
                        class="instance MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                        data-colindex="2"
                        data-field="instance"
                        role="cell"
                        style="min-width: 0; max-width: 0; min-height: auto; max-height: none;"
                        tabindex="-1"
                      >
                        <div
                          style="margin: 5px 0px 5px 0px; display: flex;"
                        >
                          <span
                            style="margin-right: 10px;"
                          >
                            MaxCut
                            :
                          </span>
                          <div
                            style="margin-right: 10px;"
                          >
                            <a
                              class="css-1wumwsl"
                              href="/classes/MaxCut/instances/j_instance"
                            >
                              j_instance
                            </a>
                          </div>
                          <div
                            style="margin-right: 10px;"
                          >
                            <a
                              class="css-1d1cfkq"
                              href="/classes/MaxCut/instances/k_instance"
                            >
                              k_instance
                            </a>
                          </div>
                          <div
                            style="margin-right: 10px;"
                          >
                            <a
                              class="css-1s7e1mx"
                              href="/classes/MaxCut/instances/l_instance"
                            >
                              l_instance
                            </a>
                          </div>
                        </div>
                        <div
                          style="margin: 5px 0px 5px 0px; display: flex;"
                        >
                          <span
                            style="margin-right: 10px;"
                          >
                            Tsp
                            :
                          </span>
                          <div
                            style="margin-right: 10px;"
                          >
                            <a
                              class="css-1i2dnn3"
                              href="/classes/Tsp/instances/g_instance"
                            >
                              g_instance
                            </a>
                          </div>
                          <div
                            style="margin-right: 10px;"
                          >
                            <a
                              class="css-c8py96"
                              href="/classes/Tsp/instances/h_instance"
                            >
                              h_instance
                            </a>
                          </div>
                          <div
                            style="margin-right: 10px;"
                          >
                            <a
                              class="css-10okgg3"
                              href="/classes/Tsp/instances/i_instance"
                            >
                              i_instance
                            </a>
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      aria-rowindex="4"
                      aria-selected="false"
                      class="MuiDataGrid-row MuiDataGrid-row--lastVisible MuiDataGrid-row--dynamicHeight"
                      data-id="d_client_unknown"
                      data-rowindex="2"
                      role="row"
                      style="max-height: none; min-height: auto;"
                    >
                      <div
                        aria-colindex="1"
                        aria-colspan="1"
                        class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                        data-colindex="0"
                        data-field="client"
                        role="cell"
                        style="min-width: 140px; max-width: 140px; min-height: auto; max-height: none;"
                        tabindex="-1"
                      >
                        <a
                          href="/clients/d_client"
                        >
                          d_client
                        </a>
                      </div>
                      <div
                        aria-colindex="2"
                        aria-colspan="1"
                        class="MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                        data-colindex="1"
                        data-field="version"
                        role="cell"
                        style="min-width: 220px; max-width: 220px; min-height: auto; max-height: none;"
                        tabindex="-1"
                      >
                        <div
                          class="MuiDataGrid-cellContent"
                        >
                          unknown
                        </div>
                      </div>
                      <div
                        aria-colindex="3"
                        aria-colspan="1"
                        class="instance MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                        data-colindex="2"
                        data-field="instance"
                        role="cell"
                        style="min-width: 0; max-width: 0; min-height: auto; max-height: none;"
                        tabindex="-1"
                      >
                        <div
                          style="margin: 5px 0px 5px 0px; display: flex;"
                        >
                          <span
                            style="margin-right: 10px;"
                          >
                            Cvrp
                            :
                          </span>
                          <div
                            style="margin-right: 10px;"
                          >
                            <a
                              class="css-zzcu01"
                              href="/classes/Cvrp/instances/o_instance"
                            >
                              o_instance
                            </a>
                          </div>
                          <div
                            style="margin-right: 10px;"
                          >
                            <a
                              class="css-2ijomo"
                              href="/classes/Cvrp/instances/p_instance"
                            >
                              p_instance
                            </a>
                          </div>
                          <div
                            style="margin-right: 10px;"
                          >
                            <a
                              class="css-19sw7en"
                              href="/classes/Cvrp/instances/q_instance"
                            >
                              q_instance
                            </a>
                          </div>
                        </div>
                        <div
                          style="margin: 5px 0px 5px 0px; display: flex;"
                        >
                          <span
                            style="margin-right: 10px;"
                          >
                            Qap
                            :
                          </span>
                          <div
                            style="margin-right: 10px;"
                          >
                            <a
                              class="css-11vui9t"
                              href="/classes/Qap/instances/m_instance"
                            >
                              m_instance
                            </a>
                          </div>
                          <div
                            style="margin-right: 10px;"
                          >
                            <a
                              class="css-1jpc6yz"
                              href="/classes/Qap/instances/n_instance"
                            >
                              n_instance
                            </a>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="Mui-resizeTriggers"
            >
              <div
                class="expand-trigger"
              >
                <div
                  style="width: 1px; height: 1px;"
                />
              </div>
              <div
                class="contract-trigger"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
